Refactor Model layers

paikwiki opened the pull request:
Close #5

paikwiki committed:
Refactor Answer class

paikwiki committed:
Remove a custom type: AnswerNumbers

paikwiki committed:
Refactor pitches member variable of Inning class

- Rename pitches to pitching
- Change type from Charactor to Int

paikwiki committed:
Rename a member variable of ViewController

- Rename labels to inningViews

paikwiki committed:
Move inningCount from Inning to Game class

paikwiki committed:
Change Inning type from Class to Struct

paikwiki committed:
Move guard statements from ViewController to Game

paikwiki committed:
Create a private method: showResult(finalResult:Bool)

paikwiki committed:
Apply CustomStringConvertible protocol to Inning class

- Rename a variable of Inning: pitchesString to description

paikwiki committed:
Move codes for judgement of threeStrikes from ViewController to Game

rinndash commented on NumberBaseball/Answer.swift:
```diff
@@ -6,36 +6,27 @@
-        var answerSet = Set<Character>()
-        while answerSet.count < 3 {
-            answerSet.insert( Character(String(Int(arc4random_uniform(10)))) )
+        var numberSet = Set<Int>()
```
숫자 야구의 도메인을 생각해 봤을 때 숫자가 3개 이상으로 늘어날 일도 없으니 숫자를 랜덤으로 생성하는데 반복문까지 필요할까 싶어. 랜덤 함수를 세 번 호출해주는쪽이 낫지 않을까?

코드도 6줄에서 3줄로 줄고 `var` 변수도 안쓰고 `popFirst()`를 하면서 `!`를 안 붙여도 되니깐. 물론 내 생각이 정답은 아니고 취향의 영역으로 볼 수도 있으니 함 생각해 봐.

```swift
first = Int(arc4random_uniform(10))
second = Int(arc4random_uniform(10))
third = Int(arc4random_uniform(10))
```

paikwiki replied:
중복되지 않는 3개의 숫자를 구해야하기 때문에 이런 방법을 썼습니다. 
안 그래도 `strikeCount`와 `ballCount`를 계산하는 코드 부분과 이 부분이 뭔가 개선해보고 싶은 부분이긴 한데...좀더 나은 방법을 찾아볼게.

rinndash replied:
아하 중복 케이스를 막아야 되는 건 생각 못했네 ㅇㅋㅇㅋ

paikwiki replied:
`Answer`의 초기화 부분을 이렇게 바꿔봤습니다.

```swift    
init() {
        var numberSet: Set<Int> = Set<Int>()
        while numberSet.count < 3 {
            numberSet.insert(Int(arc4random_uniform(10)))
        }
        let answerNumbers: [Int] = Array(numberSet)
        (first, second, third) = (answerNumbers[0], answerNumbers[1], answerNumbers[2])
        print(self.description)
    }
```

`numberSet`은 무조건 3개의 정수를 가지므로, `numberSet`을 배열로 변환한 후 `answerNumbers`에 담아서 처리하는 코드로 바꿨어. 근데 `(firts, second, third)`에 `answerNumbers`의 요소들을 담을 때, 뭔가 누적해서 처리하는, 자바스크립트의 `Array.prototype.reduce()`같은 게 있지 않을까 하고 이것저것 해보다가 결국 포기하고 지금의 코드로 남겨뒀지요... 그래도 강제언래핑도 떼어내고 코드 중복도 제거했으니 나름 만족...

rinndash commented on NumberBaseball/Inning.swift:
```diff
@@ -6,36 +6,19 @@
+    var description: String {
         var pitchesCharaterArray = Array("_  _  _")
-        for index in 0..<pitches.count {
+        for index in 0..<pitching.count {
```
이 함수를 for문 대신 map을 사용하도록 한 번 수정해 봤으면 좋겠어. 그러면 위의 var 변수를 사용하지 않을 수 있음

rinndash commented on NumberBaseball/ViewController.swift:
```diff
@@ -21,14 +21,12 @@ class ViewController: UIViewController {
     @IBOutlet var inningView08: InningView!
     @IBOutlet var inningView09: InningView!
```
`weak var`는 해당 참조의 참조 카운팅을 올리지 않지만 그냥 `var`는 참조 카운팅을 1 올립니다. 해당 객체는 스토리도드에서 생성해 스토리보드가 참조를 가지고 있기 때문에 이미 참조 카운트가 1인 상황이고요. 

여기서 `var`를 쓴다고 해서 당장 문제가 생기는 것은 아니지만, 추후 순환 참조 문제로 이어져서 메모리 누수가 발생할 수 있는 가능성이 생기고, XCode에서 스토리보드를 끌어올 때 `weak var`로 생성해 주는 것이 디폴트이기 때문에 iOS 개발의 컨벤션이라고 볼 수 있습니다.

특별한 이유가 있는게 아니면 스토리보드에서 생성한 인스턴스의 참조에는 `weak`를 붙여주도록 합시다.

rinndash commented on NumberBaseball/ViewController.swift:
```diff
@@ -21,14 +21,12 @@ class ViewController: UIViewController {
         super.viewDidLoad()
-        labels = [
+        inningViews = [
             1: inningView01!,
```
선언시 `!`를 붙인 참조는 사용할 때 `!`를 안 붙여도 쓸 수 있습니다

rinndash commented on NumberBaseball/Game.swift:
```diff
@@ -8,47 +8,74 @@
-    private let inning: Inning
+    private(set) var answer: Answer
+    private(set) var inning: Inning
     private(set) var totalInning = 9
```
나는 클래스나 구조체의 프로퍼티를 선언할 때에는 타입을 명시해주는게 좋다고 생각해. 스위프트의 타입 추론이 편하긴 한데, 몇 가지 비용이 있음. 

1. 컴파일 시간이 더 걸림 
    - 타입 추론이라는 게 일종의 방정식을 푸는 거기 때문에 컴파일 시간이 더 걸림
    - 작은 프로젝트에서는 큰 문제 아니지만 큰 프로젝트에서는 실질적인 차이가 있을 수 있음

2. 스위프트는 타입이 아주 강한 언어기 때문에 타입을 명시해 주는 버릇을 들이는게 타입 감각을 기를 수 있음
    - 타입 추론을 쓰면 자바스크립트나 파이썬 같은 언어처럼 타입을 명시 안하고 쓸 수 있고 그게 편할 수 있지만, Protocol Associated Type이나 제네릭 같은 Swift의 문법을 사용할 때 타입 개념이 없으면 컴파일조차 안됨. 그래서 가능하면 형이 프로퍼티를 선언할 때 타입을 명시해주는 버릇을 들이는 게 강한 타입 언어에 익숙해 지는데 도움이 된다고 생각해

나는 함수에서 쓰는 로컬 변수까지 타입을 명시하는 편이긴 한데, 그 정도까진 개취일 것 같고, 최소한 클래스나 구조체에서 사용하는 저장 프로퍼티는 타입을 명시해 주는 쪽이 좋다고 생각합니다. 

SwiftLint에 구조체나 클래스의 탑클래스 프로퍼티에 타입을 명시하는 것을 강제하는 룰이 있으니 그걸 추가하고 코드의 다른 부분도 적용해 봤으면 좋겠어

rinndash commented on NumberBaseball/Game.swift:
```diff
@@ -8,47 +8,74 @@
-                } else {
-                    ballCount += 1
-                }
+        if inning.pitching.firstIndex(of: answer.first) != nil {
```
여기서 이닝 결과를 문자열로 바꾸는 함수가 있는데, 두 개 정도 수정해 보면 좋을 것 같아

1. `strikeCount`와 `ballCount`를 계산하는 코드를 간결하게 수정하기
지금 비슷한 코드가 세 번이나 작성이 되어서 중복이 많은 거 같아. 이 코드를 어떻게 수정할 수 있을지 생각해 봐. 힌트는 answer의 숫자 세 개를 배열로 만들고 전에 잠깐 보여줬던 `zip` 함수를 이용해 봐

```swift
let pitches: [Int] = inning.pitching
let answerNumbers: [Int] = [answer.first, answer.second, answer.third]
let zipped: [(Int, Int)] = zip(pitches, answerNumbers)
```

2. 이 함수를 스트라이크/볼 카운트를 계산하는 함수와 이 결과를 원하는 문자열로 변경하는 함수로 분리하기
이 함수를 보면 스트라이크/볼 카운트를 계산하는 함수와 그 결과를 문자열로 포맷하는 함수가 같이 있는데 이 둘을 분리했으면 좋겠어
여기서 이닝 결과를 계산하는 것은 도메인 로직이라고 볼 수 있어. 그리고 이 결과를 라벨에 표시하기 위해 문자열로 변경하는 것은 뷰 로직이라고 할 수 있고. 그러니까 이 함수는 지금 도메인 로직과 뷰 로직이 한 곳에 있는 거지.

지금은 MVC 아키텍처로 만드는 중이라 구분이 좀 힘들 수 있지만, MVVM 패턴에서는 도메인 모델의 현재 상태를 뷰에 그리기 위한 포맷으로 변경하는 역할은 뷰모델이 해. 현재 업계에서는 MVVM 패턴이 주류라고 볼 수 있으니까 나중에 MVVM 패턴으로 개발하기 전에 모델 로직과 뷰 로직을 구분하는 연습을 해 두면 도움이 될 듯

그리고 스트라이크/볼 카운트를 계산하는 도메인 로직을 분리한다고 했을 때 분리한 함수가 어디에 있는 게 적절할지도 한 번 고민해 봐. 그게 `Game`, `Inning`, `Answer` 중 어떤 곳에 있는게 가장 적절할까?

paikwiki replied:
이 수정사항에 대해 엄청 고민해봤는데, 처음에는 `Answer`가 판단해야한다고 생각했거든. 그런데 `Answer`가 답을 비교해서 스트라이크인지 볼인지 알려준다는 게 어딘지 어색해서 결국 `Umpire`라는 클래스를 만들고 이 클래스가 판정하도록 했습니다. `Judge`라고 하려다가 야구에서는 심판을 `Umpire`라고 한다 하여 익숙치 않은 단어를 선택했지.

막상 `Umpire`를 만들고 보니 `Game`이 너무 많은 곳에 개입하고 있는 것 같아서, 이를 줄여보려고 `Umpire`만 `Answer`를 알고 있도록 변경했습니다. 이렇게 하니까 `Game`의 변수 중 `isThreeStrike`가 여기저기에 너무 많이 쓰이고 있다는 걸 알게 됐고, 이유를 생각해보니 게임을 끝내야하는지 알기 위해서 그랬더군.

이를 해결하기 위해 `isGameFinish` 같은 변수를 만들까 하다가, 그보다는 게임의 결과를 담당할 객체를 만들기로 하고 `GameResult`를 만들었어. `win`과 `lose`밖에 없지만 이렇게 하는 게 더 가독성이 좋을 것 같아서 한번 해봤습니다.

`zip()` 관련해서는 어떻게 쓰는 게 좋을지 고민하다가 `Zip2Sequence<[Int], [Int]>`라는 타입을 잘 다루기가 어려워서 `answerNumbers`라는 배열에 세 자리를 담아서 리팩토링해봤습니다. `(0..<N).forEach()` 문법이 너무 편리하네.

내가 작업한 코드에서 많은 부분들이 아직 의문이지만.. 아래 부분을 한번 봐주겠습니까?

```swift
struct Umpire: CustomStringConvertible {
    private let answer: Answer
    var description: String {
        answer.description
    }
    ...

}
```

경기가 끝나고 나면 `Umpire`가 화면에 정답을 출력하기 위해서 `CustomStringConvertible` 프로토콜을 따르도록 했는데, `umpire.description`을 부르면 그 안에서 ` answer.description`을 부르도록 해뒀거든. 그런데 `description`이라면 뭔가 자기 자신에 대해 서술해야 할 것 같은데, 이건 `Answer`에 대한 서술이니까.. 이런 경우엔 `description`보다는 다른 이름이 좋을까?

이 리뷰에 대해 수정하면서 이것저것 너무 많이 한꺼번에 고쳐버려서 하나의 커밋에 너무 많은 수정사항이 들어갔네요. 이거.. 분리할 엄두가 안나서 커밋을 한방에 했습니다. 앞으로 주의해서 잘 분리해두겠습니다.

이와 관련한 커밋은 [d8a3161](https://github.com/paikwiki/ios-number-baseball/pull/11/commits/d8a3161941dd19e14763151804c9601e62660901): "Create Umpire class and GameResult enum" 입니다.

rinndash replied:
Umpire 타입에 코멘트 달았습니다

rinndash commented on NumberBaseball/Game.swift:
```diff
@@ -8,47 +8,74 @@
 
-    func isThreeStrikes() -> Bool {
-        inning.pitches == answer.answer
+    func getNextInning() {
```
함수 이름이 적절하지 않은듯 일반적으로 `get`은 `getter` 함수에 쓰는 접두사기 때문에 오해의 소지가 있다고 생각해. 가능하면 도메인에 쓰는 용어를 쓰는게 좋으니 이닝을 끝낸다, 이닝을 마무리한다고 흔히 표현하니 `func endInning()` 같은 이름이 더 적절하지 않을까 싶어. 영어로도 end inning 이런 표현을 쓰는지는 잘 모르겠으니 함 찾아보는게 좋을 듯

rinndash commented on NumberBaseball/Inning.swift:
```diff
@@ -6,36 +6,19 @@
-    private(set) var pitches = [Character]()
-    var pitchesString: String {
+    var pitching = [Int]()
+    var isThrowThreeBalls: Bool {
```
이 프로퍼티의 이름은 적절하지 않은 거 같아. `isThrowThreeBalls` 이라는 이름은 공을 던졌을 때 스트라이크가 아닌 볼만 세개를 던졌는가로 오해하기 쉬운 것 같아. 공 세 개를 던져서 이닝이 끝났는지를 표현하기 위해선 `isEnded` 같은 이름이 적절하지 않을까

rinndash reviewed (changes requested):
코멘트 남긴 부분 수정 바랍니다~

paikwiki committed:
Change types of IBOutlet from strong to weak

paikwiki committed:
Remove forced unwrapping for inningViews

paikwiki committed:
Refactor description variable of Inning class

paikwiki committed:
Rename a variable of Inning class

- isThrowThreeBalls -> isEnded

paikwiki committed:
Rename a variable of Inning class

- isThrowThreeBalls -> endInning

paikwiki committed:
Resolve conflicts

paikwiki committed:
Rename a method of Inning  class

- getNextInning() -> startNextInning()

paikwiki committed:
Create .swiftlint.yml

- Add rule: explicit_type_interface

paikwiki committed:
Lint codes with Swiftlint rule: explicit_type_interface

rinndash commented on NumberBaseball/ViewController.swift:
```diff
@@ -10,73 +10,73 @@ import UIKit
-    private let answer = Answer()
-    private let inning = Inning()
-    lazy private var game = Game(answer: answer, inning: inning)
+    private var inningViews: [Int: InningView] = [Int: InningView]()
```
이제 `inningViews`가 딕셔너리일 필요가 없는 거 같음. 0부터 9까지 배열로도 충분히 필요한 이닝을 가져올 수 있을 것 같아

paikwiki replied:
배열로 바꾸면서 보니까 `didTapNumber(_ sender:)`에서 

```swift
inningViews[game.inningCount]?.pitchesLabel.text = game.inning.description
```

이렇게 옵셔널 붙여준 덕분에 에러가 안 나고 있었을 뿐, 숫자패드를 누를 때마다 레이블을 업데이트하려는 시도를 하고 있었네... 이 부분도 수정했음. 하지만 `game.isOver`를 썼기 때문에 또 고쳐야 할듯... 아래 수정사항 고치면서 함께 손볼 생각입니다.

rinndash commented on NumberBaseball/ViewController.swift:
```diff
@@ -10,73 +10,73 @@ import UIKit
-            inning.pitches.contains(pitchNumber) == false,
-            game.isOver == false
+    @IBAction func didTapNumber(_ sender: UIButton) {
+        guard let pitchNumber = Int(sender.currentTitle!)
```
`!` 쓰지 않도록 해주세요. SwiftLint 보면 `!` 사용 자체를 막는 룰이 있으니 그걸 추가하는 게 좋을 듯

rinndash commented on NumberBaseball/ViewController.swift:
```diff
@@ -10,73 +10,73 @@ import UIKit
         // UI UPDATE
         answerLabel.text = "X   X   X"
-        labels.forEach({ inning in
+        inningViews.forEach({ inningView in
```
inningViews를 배열로 변경하면 필요없는 이야기일 수도 있는데 이렇게 튜플을 클로저의 인풋 타입으로 받을 때에는 `in` 키워드 앞에 튜플의 파라미터를 두 개 받는 쪽이 더 낫다고 생각해.

```swift
inningView.forEach { key, value in
  // blah blah
}
```

이런 경우에는 key에는 접근하지 않으니까 `_`로 받으면 되고
```swift
inningView.forEach { _, inningView in
  // blah blah
}
```

그리고 가능하면 trailing closure를 사용해 주세요. 역시 SwiftLint에 트레일링 클로저 관련 룰이 있을텐데 그것도 추가해주세요

rinndash commented on NumberBaseball/Game.swift:
```diff
@@ -8,47 +8,74 @@
+    private(set) var answer: Answer
+    private(set) var inning: Inning
+    private(set) var totalInning: Int = 9
+    private(set) var isOver: Bool  = false
```
이 프로퍼티는 저장 프로퍼티일 필요가 없다고 생각해. 현재 게임이 끝났는지는 `inningCount`와 `inning`의 `isEnded`값, 그리고 `isThreeStrikes` 값으로 계산할 수 있는 프로퍼티인 듯

rinndash commented on NumberBaseball/Game.swift:
```diff
@@ -8,47 +8,74 @@
+        inning = Inning()
     }
 
     func gameOver() {
```
`isOver`가 계산 프로퍼티이면 이 함수는 필요가 없겠지?

paikwiki committed:
Create Umpire class and GameResult enum

- Detach logic for judgement of each inning and generating game result from Game
- Game cant know Answer, only Umpire know it.
- GameResult has two cases: .win and .lose
- ViewController can make a decision whether hide or show answer with GameResult
- Game does not need to have a variable "isThreeStrikes" because of GameResult

paikwiki committed:
Add a rule on .swiftlint: force_unwrapping

paikwiki committed:
Rearrange items of a project navigator

paikwiki committed:
Lint codes with Swiftlint rule: force_unwrapping

paikwiki committed:
Add a rule on .swiftlint: trailing_closure

paikwiki committed:
Lint codes with Swiftlint rule: trailing_closure

paikwiki committed:
Change type of inningViews on ViewController: Dictionary->Array

paikwiki committed:
Change a property type of isOver on Game: Stored->Computed

paikwiki commented:
> 코멘트 남긴 부분 수정 바랍니다~

수정 완료했습니다. 뷰 하나짜리 앱에서도 이렇게 많은 부분을 수정해야 한다니 앞길이 깜깜-하네요!! :see_no_evil: :see_no_evil: :see_no_evil:

rinndash commented on NumberBaseball/Game.swift:
```diff
@@ -8,47 +8,54 @@
+        "\(inningResult.strikeCount)S \(inningResult.ballCount)B"
+    }
+    var isOver: Bool {
+         inningResult.strikeCount == 3 || (inningCount == totalInning && inning.pitching.count == 3)
```
`inning.pitching.count==3`보다는 `inning.isEnded`가 낫지 않을까

rinndash commented on NumberBaseball/Game.swift:
```diff
@@ -8,47 +8,54 @@
-    func gameOver() {
-        isOver = true
-        print("GAME OVER")
+    func showAnswer() {
```
필요없는 메소드는 지워주세요

paikwiki committed:
Refactor if statement for Computed property(isOver) of Game class

paikwiki committed:
Remove useless method: showAnswer() of Game class

rinndash commented on NumberBaseball/Game.swift:
```diff
@@ -8,47 +8,54 @@
+        return
     }
 
     func reset() {
```
개인적으로는 `Game`을 리셋하는 것보다는 새 `Game`을 생성하는 게 좋다고 생각해. 이유는 Game에 상태가 추가됐을 때 reset() 코드도 같이 변경해줘야 되기 때문에 관리 포인트가 늘어나고 그만큼 실수할 여지도 더 생김.

그리고 숫자 야구에 `Game`을 리셋한다는 개념은 없잖아? 그냥 새 게임을 시작하는 거일 뿐이지. 다시 말해 도메인 모델에 재활용하는 개념은 없는 상황인거야

그렇다면 왜 `reset()`를 쓰는 걸까? 내 생각에는 한 번 메모리에 올린 `Game`의 인스턴스를 재활용하기 위해서지. 이미 메모리에 올라간 인스턴스를 재활용하는게 새로 인스턴스를 생성하는 것보다 컴퓨팅 비용이 낮아. 즉 도메인 모델에 재활용이라는 개념이 없다면 `reset()` 코드는 성능을 최적화하기 위한 코드인거지. 

iOS를 개발할 때 이런식으로 기존에 이미 생성된 인스턴스를 재활용하는 상황은 `UITableViewCell`이나 `UICollectionViewCell`처럼 화면에는 소수의 인스턴스만 필요한데 표현해야할 도메인 모델의 숫자는 많은 경우 화면에 보이는 인스턴스만 재활용할 경우, 메모리도 아끼고 시간도 아낄 수 있지.

그렇다면 현재 상황에서 Game의 인스턴스를 재활용할 필요가 있을까? 나는 없다고 생각해. Game 타입 뿐만이 아니라 앞으로 어떤 모델을 작성할 때 이게 재활용이 필요한 것인지 생각해보고 새 인스턴스를 생성해서 쓰는 쪽이 성능상 큰 비용이 아니라면 그냥 새로 만들어서 쓰도록 구현하는게 유지보수 비용을 낮추고 코드를 간결하게 쓸 수 있어

rinndash commented on NumberBaseball/GameResult.swift:
```diff
@@ -0,0 +1,12 @@
+//  Copyright © 2020 Changhyun Baek. All rights reserved.
+//
+
+enum GameResult {
```
이렇게 `GameResult`에 case만 2개 있다면 난 굳이 별도의 타입으로 만들지 않고 Bool을 써도 될 것 같아. 다른 타입에서 게임 결과와 관련된 프로퍼티나 함수들을 찾아보고 있다면 이 타입으로 옮겨봐~

rinndash commented on NumberBaseball/Umpire.swift:
```diff
@@ -0,0 +1,35 @@
+//  Copyright © 2020 Changhyun Baek. All rights reserved.
+//
+
+struct Umpire: CustomStringConvertible {
```
개인적으로는 심판이라는 모델까지 필요한 것 같지는 않아. 이 애플리케이션의 도메인은 `숫자 야구`지 `야구`는 아니니까. 숫자 야구에 굳이 심판까지 필요한 것 같지는 않거든. 그런데 좋은 시도라고 생각해.

하지만 심판의 `description`이 `answer.description`을 말하는 것은 좀 아닌 거 같긴 해. 이런 경우에는 차라리 `Answer` 타입을 `internal`로 노출하는 게 나을 것 같기도 하고. 근데 그렇게 되면 굳이 `Game`에서 `Umpire`를 따로 분리할 필요도 없을 거 같기도 하고..

도메인 모델링에는 정답은 없으니깐 형이 충분히 생각해보고 수정이 필요하다면 수정해봐~

rinndash commented:
이번 PR에서는 이번에 단 수정 사항까지만 적용하고 머지해도 괜찮을 거 같아. 그래도 처음보다는 많이 코드가 깔끔해 진 것 같으니 힘을 내십쇼! 오늘 중에 디스코드 한번 합시다~

paikwiki committed:
Remove Umpire class and replace it with Answer class

paikwiki committed:
Remove GameResult Enum and replace it with a Boolean variable

paikwiki committed:
Recreate Game instance when the game is restarted

paikwiki commented:
> 이번 PR에서는 이번에 단 수정 사항까지만 적용하고 머지해도 괜찮을 거 같아. 그래도 처음보다는 많이 코드가 깔끔해 진 것 같으니 힘을 내십쇼! 오늘 중에 디스코드 한번 합시다~

이제 코드베이스를 많이 다듬어서 초반에 비해 리팩토링 작업이 많이 편해졌네요ㅎㅎ 오후 일정 마치고나면 연락드리겠습니다.

rinndash reviewed (approved):
고생 많았습니다~
👍 💯 🎉
